doc/design.md

```diff
@@ -0,0 +1,1268 @@
+// be the (only) dimension of the nested views.
+DatasetView<Histogram, const Coord::SpectrumNumber>
+    view(lambdaSlice, "sample - background", {Dim::Temperature});
+
+for (const auto &tempDependence : view) {
+  // Items in view have named getters, alternatively
+  // get<Coord::SpectrumNumber>() can be used.
+  printf("Plotting spectrum %d\n", tempDependence.spectrumNumber());
+  // Plot the data.
+  mantid::matplotlib::plot(tempDependence.get<Histogram>());
+  // The view we are passing to `plot` can be operated on as follows.
+  for (const auto bin : tempDependence.get<Histogram>()) {
+    // Note that we can read left, right, and center of the bin, even though
+    // the underlying data is storing the edges.
+    mantid::matplotlib::drawBox(bin.left(), bin.right(), bin.value());
+    // Note that we can read the standard deviation, even though the
+    // underlying data is stored as variance.
+    mantid::matplotlib::drawErrorBar(bin.center(), bin.value(), bin.stdDev());
+  }
+}
+```
+
+
+### <a name="examples-python"></a>Python example
+
+The design of the Python API is in an early stage so we do not give a full example yet.
+Overall the functionality will be as above for the C++ example.
+If in doubt regarding conventions, we intend to adopt the behavior of `numpy` (or `xarray`, if applicable), in particular for indexing and slicing.
+
+
+## <a name="design-details"></a>Design details
+
+
+### <a name="design-details-dependencies"></a>Dependencies
+
+- `boost::units` for unit handling.
+- `boost::container` for `small_vector`, which is used as an optimization in a couple of places.
+- `boost::mpl` to help with some meta programming.
+- `Eigen` for vectors and rotations, in particular for detector positions.
+- `gsl` is used for `gsl::index` and `gsl::span`.
+  This could easily be replaced by a stand alone implementation, in particular since `span` is part of C++20.
+- `range-v3` has been used for implementing some multi-column sort operations but replacing it by a stand-alone implementation should be no trouble.
+- GoogleTest for unit testing.
+- GoogleBenchmark for benchmarking.
+
+The prototype has been implemented using C++14, and this is sufficient.
+Nevertheless, having C++17 available would be an advantage, particularly since a moderate amount of template-meta programming is required.
+Given the long time scale, it could be expected that the Mantid project is ready to move to C++17 by the time `Dataset` might be included in the main repository.
+
+Note that `Dataset` or rather `Variable` will be able to hold data from libraries that are *not* dependencies of this library.
+It is sufficient to forward-declare types when adding a tag for such a variable.
+*Example:
+We can store `API::Run` in `Dataset`, but the dataset library does not need to depend on `Mantid::API`.*
+
+It is currently unclear whether this library should be kept as a separate repository or be included as a new module in the Mantid repository.
```

Comment:
So if we kept it external we'd probably use the external data mechanism to bring it in?

Reply:
I had in mind `cmake`'s `ExternalProject`.

Reply:
Sorry, I mistyped, I also meant the `ExternalProject` so that's fine.